src/main/java/at/hannibal2/skyhanni/data/PetAPI.kt

```diff
@@ -35,31 +193,425 @@ object PetAPI {
         set(value) {
             ProfileStorageData.profileSpecific?.currentPet = value
         }
 
-    fun isCurrentPet(petName: String): Boolean = currentPet?.contains(petName) ?: false
+    fun isCurrentPet(petName: String): Boolean = pet?.cleanName?.contains(petName) ?: false
 
     fun getCleanName(nameWithLevel: String): String? {
-        petItemName.matchMatcher(nameWithLevel) {
+        petItemNamePattern.matchMatcher(nameWithLevel) {
             return group("name")
         }
-        neuRepoPetItemName.matchMatcher(nameWithLevel) {
+        neuRepoPetItemNamePattern.matchMatcher(nameWithLevel) {
             return group("name")
         }
 
         return null
     }
 
-    fun getPetLevel(nameWithLevel: String): Int? = petItemName.matchMatcher(nameWithLevel) {
+    @Deprecated(message = "use PetAPI.pet.level")
+    fun getPetLevel(nameWithLevel: String): Int? = petItemNamePattern.matchMatcher(nameWithLevel) {
         group("level").toInt()
     }
 
-    fun hasPetName(name: String): Boolean = petItemName.matches(name) && !ignoredPetStrings.any { name.contains(it) }
+    @Deprecated(message = "use PetAPI.pet.name")
+    fun hasPetName(name: String): Boolean = petItemNamePattern.matches(name) && !ignoredPetStrings.any { name.contains(it) }
+
+// ---
+    @SubscribeEvent
+    fun onWidgetUpdate(event: WidgetUpdateEvent) {
+        if (!event.isWidget(TabWidget.PET)) return
+
+        var newPetLine: String? = null
+        for (line in event.lines) {
+            if (petWidgetPattern.matches(line)) {
+                newPetLine = line.removePrefix(" ")
+                break
+            }
+        }
+        if (newPetLine == pet?.rawPetName || newPetLine == null) return
+
+        var petItem = NEUInternalName.NONE
+        var petXP: Double? = null
+        event.lines.forEach { line ->
+            val tempPetItem = handleWidgetStringLine(line)
+            if (tempPetItem != NEUInternalName.NONE) {
+                petItem = tempPetItem
+                return@forEach
+            }
+            val tempPetXP = handleWidgetXPLine(line)
+            if (tempPetXP != null) {
+                petXP = tempPetXP
+                return@forEach
+            }
+        }
+
+        event.lines.forEach { line ->
+            if (handleWidgetPetLine(line, petItem, petXP, newPetLine)) return@forEach
+        }
+    }
+
+    private fun handleWidgetPetLine(line: String, petItem: NEUInternalName, petXP: Double?, newPetLine: String): Boolean {
+        val xpOverLevel = petXP ?: 0.0
+        petWidgetPattern.matchMatcher(line) {
+            val xp = (levelToXP(
+                group("level").toInt(),
+                LorenzRarity.getByColorCode(group("rarity")[0]) ?: LorenzRarity.ULTIMATE,
+                group("name")
+            ))
+            val petName = group("name")
+            val rarity = LorenzRarity.getByColorCode(group("rarity")[0]) ?: LorenzRarity.ULTIMATE
```

Comment:
you are getting the same groups multiple times in this matcher, make vals of them so you only have to access them once